documentation/jetty-documentation/src/main/asciidoc/programming-guide/websocket.adoc

```diff
@@ -248,18 +221,11 @@ While non-blocking APIs are more difficult to use, they don't block the sender t
 
 If you need to send large WebSocket messages, you may reduce the memory usage by streaming the message content.
 
-Both blocking and non-blocking APIs offer `sendPartial*(...)` methods that allow you to send a chunk of the whole message at a time, therefore reducing the memory usage since it is not necessary to have the whole message `String` or `byte[]` in memory to send it.
-
-Streaming sends using blocking APIs is quite simple:
-
-[source,java,indent=0]
-----
-include::{doc_code}/org/eclipse/jetty/docs/programming/WebSocketDocs.java[tags=streamSendBlocking]
-----
+The APIs offer `sendPartial*(\...)` methods that allow you to send a chunk of the whole message at a time, therefore reducing the memory usage since it is not necessary to have the whole message `String` or `ByteBuffer` in memory to send it.
 
-Streaming sends using non-blocking APIs is more complicated, as you should wait (without blocking!) for the callbacks to complete.
+The APIs for streaming the message content are non-blocking and therefore slightly more complicated to use, as you should wait (without blocking!) for the callbacks to complete.
 
-Fortunately, Jetty provides you with the `IteratingCallback` utility class (described in more details xref:pg-arch-io-echo[in this section]) which greatly simplify the use of non-blocking APIs:
+Fortunately, Jetty provides the `IteratingCallback` utility class (described in more details xref:pg-arch-io-echo[in this section]) which greatly simplify the use of non-blocking APIs:
```

Comment:
The only way to avoid users needing to do something like `IteratingCallback` is to have a pull API like our `Content.Source` is.   That way, the user can iterate calling something like `readFrame()` until they get a null return and then they can do a `demand` to get an `onFrameAvailable()`.     With such an API, the infinite recursion problem of `demand` calling `onFrame` calling `demand` calling `onFrame` etc. is removed.

Hhmmmm but we could also use the `SerializedInvoker` to ensure that if `onFrame` calls `demand`, that we do not invoke `onFrame` again until the previous one returns? Do we do that?  If so, then we don't need to promote the `IteratingCallback`.  Let me go read what the contract actually is.

Reply:
javadoc on FrameHandler says 
>     * <p>This method will never be called concurrently for the
>     * same session

So users don't have infinite recursion problems.   Probably don't need to push `IteratingCallback` to them.

Reply:
@joakime we have a whole section in the docs about `IteratingCallback` so it is documented to be used properly.

The problem is not really reading asynchronously, it's writing asynchronously.
You read one chunk of data, you need to write it passing it a callback.
Then you have to "wait" until the callback is succeeded, so you cannot loop, etc., you really need `IteratingCallback`, or some other APIs (that we don't have for WebSocket) that has an underlying `IteratingCallback`.

I'm all for new APIs, but can be done in a different PR if necessary.